File: odoo_surface_mcp/odoo_client.py
"""Odoo JSON-RPC client — session-based wrapper around /web/dataset/call_kw."""
import itertools
import os
import time
from typing import Any

import requests


class OdooClient:

    def __init__(self, url: str, db: str, username: str, password: str):
        self.url = url.rstrip("/")
        self.db = db
        self.username = username
        self.password = password
        self._uid: int | None = None
        self._id_iter = itertools.count(1)
        self._session = requests.Session()
        self._session.headers.update({"Content-Type": "application/json"})

    @classmethod
    def from_env(cls) -> "OdooClient":
        # Accept both ODOO_USERNAME and ODOO_USER for backwards compatibility.
        username = (
            os.getenv("ODOO_USERNAME")
            or os.getenv("ODOO_USER")
            or "admin"
        )
        return cls(
            url=os.getenv("ODOO_URL", "http://localhost:8069"),
            db=os.getenv("ODOO_DB", "odoo17"),
            username=username,
            password=os.getenv("ODOO_PASSWORD", "admin"),
        )

    def _rpc(self, route: str, params: dict) -> Any:
        """POST a JSON-RPC 2.0 envelope to *route* and return the result.

        The session cookie is managed automatically by requests.Session, so
        the Odoo server-side session (and request.website on frontend routes)
        persists across calls within the same OdooClient instance.

        Redirects are followed manually with POST preserved on every hop.
        requests' default behaviour downgrades POST→GET on 301/302, which
        drops the JSON body and causes Odoo to return "Access Denied".
        """
        payload = {
            "jsonrpc": "2.0",
            "method": "call",
            "id": next(self._id_iter),
            "params": params,
        }
        url = f"{self.url}{route}"
        for _ in range(5):
            resp = self._session.post(url, json=payload, timeout=30, allow_redirects=False)
            if not resp.is_redirect:
                break
            location = resp.headers.get("Location", "")
            if not location:
                break
            # Absolute redirect → update self.url so future calls skip the hop.
            if location.startswith(("http://", "https://")) and route in location:
                self.url = location[: location.rindex(route)]
            url = location if location.startswith("http") else f"{self.url}{location}"
        resp.raise_for_status()
        data = resp.json()
        if "error" in data:
            err = data["error"]
            msg = (
                err.get("data", {}).get("message")
                or err.get("message")
                or str(err)
            )
            raise Exception(msg)
        return data.get("result")

    @property
    def uid(self) -> int:
        if self._uid is None:
            result = self._rpc("/web/session/authenticate", {
                "db": self.db,
                "login": self.username,
                "password": self.password,
            })
            uid = (result or {}).get("uid")
            if not uid:
                raise ConnectionError(
                    f"Odoo authentication failed: {self.username}@{self.db}"
                )
            self._uid = uid
        return self._uid

    def execute(self, model: str, method: str, *args: Any, **kwargs: Any) -> Any:
        """Call model.method(*args, **kwargs) via /web/dataset/call_kw.

        Re-authenticates once if the server-side session has expired so
        long-running MCP processes survive Odoo's session TTL without a restart.
        """
        _ = self.uid  # ensure session cookie is established
        kw = {"model": model, "method": method, "args": list(args), "kwargs": kwargs}
        try:
            return self._rpc("/web/dataset/call_kw", kw)
        except Exception as exc:
            if _is_session_expired(exc):
                self._uid = None
                _ = self.uid
                return self._rpc("/web/dataset/call_kw", kw)
            raise

    def http_call(self, route: str, params: dict | None = None) -> Any:
        """Call any Odoo JSON route within the established session.

        Covers website-specific routes (e.g. /website/snippet/filters) and
        frontend routes where request.website is bound by the website
        middleware — neither of which are reachable via execute().

        Re-authenticates once on session expiry (same as execute).
        """
        _ = self.uid  # ensure session cookie is established
        p = params or {}
        try:
            return self._rpc(route, p)
        except Exception as exc:
            if _is_session_expired(exc):
                self._uid = None
                _ = self.uid
                return self._rpc(route, p)
            raise

    def ping(self) -> dict:
        """Check connectivity and return version info with latency."""
        t0 = time.time()
        info = self._rpc("/web/webclient/version_info", {})
        rpc_ms = round((time.time() - t0) * 1000, 2)
        t1 = time.time()
        uid = self.uid
        auth_ms = round((time.time() - t1) * 1000, 2)
        return {
            "status": "ok",
            "server_version": (info or {}).get("server_version"),
            "db": self.db,
            "uid": uid,
            "rpc_latency_ms": rpc_ms,
            "auth_latency_ms": auth_ms,
        }

    def get_form_arch(self, model: str) -> str:
        """Return the compiled form view arch XML string for a model."""
        result = self.execute(model, "get_views", [[False, "form"]])
        return result["views"]["form"]["arch"]

    def get_form_fields(self, model: str) -> dict:
        """Return the fields metadata dict for a model.

        Odoo 17 get_views puts field metadata under result["models"][model],
        not inside result["views"]["form"]["fields"]. Fall back to fields_get
        if the structure is unexpected.
        """
        try:
            result = self.execute(model, "get_views", [[False, "form"]])
            # Odoo 17: field metadata is in result["models"][model]
            models_meta = result.get("models", {})
            if model in models_meta:
                return models_meta[model]
            # Older structure fallback
            form_view = result.get("views", {}).get("form", {})
            if "fields" in form_view:
                return form_view["fields"]
        except Exception:
            pass
        # Reliable fallback: fields_get returns all fields with metadata
        return self.execute(
            model, "fields_get",
            attributes=["string", "type", "relation", "required", "readonly"],
        )

    def check_access(self, model: str, operation: str) -> bool:
        """Return True if the current user can perform operation on model."""
        try:
            return bool(
                self.execute(model, "check_access_rights", operation, False)
            )
        except Exception:
            return False

    def valid_field_names(self, model: str) -> set[str]:
        """Return all valid field names on a model.

        Used to filter field lists before read() so we never request fields
        that don't exist on the model (e.g. subview fields collected from
        nested inline trees, or renamed/removed computed fields).
        """
        meta = self.execute(model, "fields_get", attributes=["string"])
        return set(meta.keys())

    def close(self) -> None:
        """Close the underlying HTTP session.  Called by the server lifespan on shutdown."""
        self._session.close()

    def get_model_id(self, model: str) -> int | None:
        """Return the ir.model.id for a model technical name, or None."""
        ids = self.execute("ir.model", "search", [["model", "=", model]])
        return ids[0] if ids else None

    def user_group_xmlids(self) -> frozenset[str]:
        """Return the set of group XML IDs the current user belongs to.

        Group XML IDs look like 'account.group_account_manager' or 'base.group_user'.
        Result is cached on the client instance for the lifetime of the session.
        """
        if hasattr(self, "_group_xmlids"):
            return self._group_xmlids  # type: ignore[attr-defined]

        # 1. Fetch group numeric IDs for the current user
        rows = self.execute("res.users", "read", [self.uid], fields=["groups_id"])
        group_ids: list[int] = rows[0].get("groups_id", []) if rows else []

        # 2. Resolve numeric IDs → XML IDs via ir.model.data
        xmlids: set[str] = set()
        if group_ids:
            imd_rows = self.execute(
                "ir.model.data", "search_read",
                [["model", "=", "res.groups"], ["res_id", "in", group_ids]],
                fields=["module", "name"],
            )
            for row in imd_rows:
                xmlids.add(f"{row['module']}.{row['name']}")

        self._group_xmlids: frozenset[str] = frozenset(xmlids)
        return self._group_xmlids


# ---------------------------------------------------------------------------
# Module-level helpers
# ---------------------------------------------------------------------------

_SESSION_EXPIRED_PHRASES = (
    "session expired",
    "session_invalid",
    "not logged",
    "odoo.http.sessionexpired",
)


def _is_session_expired(exc: Exception) -> bool:
    """Return True when *exc* signals an Odoo server-side session expiry.

    Odoo surfaces session expiry as a JSON-RPC error with messages like
    "Session Expired" or "Access Denied" (when the session cookie is stale).
    We catch those and let the caller re-authenticate rather than propagating.
    """
    msg = str(exc).lower()
    return any(phrase in msg for phrase in _SESSION_EXPIRED_PHRASES)

File: src/odooClient.ts
/**
 * Odoo JSON-RPC client — session-based wrapper around /web/dataset/call_kw.
 *
 * Calling convention for execute():
 *   execute(model, method, args, kwargs)
 *   where args  = positional arguments to the ORM method (array)
 *         kwargs = keyword arguments to the ORM method (object)
 *
 * Example:
 *   execute("res.partner", "search_read", [domain], { fields: ["name"], limit: 10 })
 */

const SESSION_EXPIRED = ['session expired', 'session_invalid', 'not logged', 'odoo.http.sessionexpired'];

function isSessionExpired(err: unknown): boolean {
  const msg = String(err).toLowerCase();
  return SESSION_EXPIRED.some(p => msg.includes(p));
}

interface RpcResponse {
  result?: unknown;
  error?: { message?: string; data?: { message?: string } };
}

export class OdooClient {
  private _uid: number | null = null;
  private _idCounter = 1;
  private _cookies = new Map<string, string>();
  private _groupXmlids: Set<string> | null = null;

  constructor(
    private url: string,
    private readonly db: string,
    private readonly username: string,
    private readonly password: string,
  ) {
    this.url = url.replace(/\/$/, '');
  }

  static fromEnv(): OdooClient {
    const url      = process.env.ODOO_URL;
    const db       = process.env.ODOO_DB;
    const username = process.env.ODOO_USERNAME ?? process.env.ODOO_USER;
    const password = process.env.ODOO_PASSWORD;
    const missing  = ['ODOO_URL', 'ODOO_DB', 'ODOO_PASSWORD'].filter(k => !process.env[k]);
    if (missing.length) throw new Error(`Missing required environment variables: ${missing.join(', ')}`);
    return new OdooClient(url!, db!, username ?? 'admin', password!);
  }

  // -------------------------------------------------------------------------
  // Cookie management
  // -------------------------------------------------------------------------

  private cookieHeader(): string {
    return [...this._cookies.entries()].map(([k, v]) => `${k}=${v}`).join('; ');
  }

  private updateCookies(response: Response): void {
    // Node 18.14+ exposes getSetCookie(); fall back to single header string.
    const setCookies: string[] =
      typeof (response.headers as unknown as { getSetCookie?: () => string[] }).getSetCookie === 'function'
        ? (response.headers as unknown as { getSetCookie: () => string[] }).getSetCookie()
        : [response.headers.get('set-cookie')].filter(Boolean) as string[];

    for (const raw of setCookies) {
      const pair = raw.split(';')[0].trim();
      const eq = pair.indexOf('=');
      if (eq < 0) continue;
      this._cookies.set(pair.slice(0, eq).trim(), pair.slice(eq + 1).trim());
    }
  }

  // -------------------------------------------------------------------------
  // Core RPC
  // -------------------------------------------------------------------------

  private async rpc(route: string, params: Record<string, unknown>): Promise<unknown> {
    const payload = { jsonrpc: '2.0', method: 'call', id: this._idCounter++, params };
    let url = `${this.url}${route}`;

    for (let hop = 0; hop < 5; hop++) {
      const cookie = this.cookieHeader();
      const headers: Record<string, string> = { 'Content-Type': 'application/json' };
      if (cookie) headers['Cookie'] = cookie;

      const resp = await fetch(url, {
        method: 'POST',
        headers,
        body: JSON.stringify(payload),
        redirect: 'manual',
      });
      this.updateCookies(resp);

      if (resp.status >= 300 && resp.status < 400) {
        const loc = resp.headers.get('location') ?? '';
        if (!loc) break;
        if ((loc.startsWith('http://') || loc.startsWith('https://')) && loc.includes(route)) {
          this.url = loc.slice(0, loc.lastIndexOf(route));
        }
        url = loc.startsWith('http') ? loc : `${this.url}${loc}`;
        continue;
      }

      if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
      const data = await resp.json() as RpcResponse;
      if (data.error) {
        const err = data.error;
        throw new Error(err.data?.message ?? err.message ?? JSON.stringify(err));
      }
      return data.result;
    }
    throw new Error(`Too many redirects for ${route}`);
  }

  // -------------------------------------------------------------------------
  // Auth
  // -------------------------------------------------------------------------

  async getUid(): Promise<number> {
    if (this._uid !== null) return this._uid;
    const result = await this.rpc('/web/session/authenticate', {
      db: this.db, login: this.username, password: this.password,
    }) as { uid?: number } | null;
    const uid = result?.uid;
    if (!uid) throw new Error(`Odoo authentication failed: ${this.username}@${this.db}`);
    this._uid = uid;
    return uid;
  }

  // -------------------------------------------------------------------------
  // ORM execute
  // -------------------------------------------------------------------------

  async execute(
    model: string,
    method: string,
    args: unknown[] = [],
    kwargs: Record<string, unknown> = {},
  ): Promise<unknown> {
    await this.getUid();
    const kw = { model, method, args, kwargs };
    try {
      return await this.rpc('/web/dataset/call_kw', kw);
    } catch (exc) {
      if (isSessionExpired(exc)) {
        this._uid = null;
        await this.getUid();
        return await this.rpc('/web/dataset/call_kw', kw);
      }
      throw exc;
    }
  }

  async httpCall(route: string, params: Record<string, unknown> = {}): Promise<unknown> {
    await this.getUid();
    try {
      return await this.rpc(route, params);
    } catch (exc) {
      if (isSessionExpired(exc)) {
        this._uid = null;
        await this.getUid();
        return await this.rpc(route, params);
      }
      throw exc;
    }
  }

  // -------------------------------------------------------------------------
  // Convenience helpers
  // -------------------------------------------------------------------------

  async ping(): Promise<Record<string, unknown>> {
    const t0 = Date.now();
    const info = await this.rpc('/web/webclient/version_info', {}) as Record<string, unknown> | null;
    const rpcMs = Date.now() - t0;
    const t1 = Date.now();
    const uid = await this.getUid();
    const authMs = Date.now() - t1;
    return {
      status: 'ok',
      server_version: info?.server_version,
      db: this.db,
      uid,
      rpc_latency_ms: rpcMs,
      auth_latency_ms: authMs,
    };
  }

  async getFormArch(model: string): Promise<string> {
    const result = await this.execute(model, 'get_views', [[[false, 'form']]]) as {
      views: { form: { arch: string } };
    };
    return result.views.form.arch;
  }

  async getFormFields(model: string): Promise<Record<string, unknown>> {
    try {
      const result = await this.execute(model, 'get_views', [[[false, 'form']]]) as {
        views: { form: { fields?: Record<string, unknown> } };
        models?: Record<string, unknown>;
      };
      const modelsMeta = result.models as Record<string, unknown> | undefined;
      if (modelsMeta?.[model]) return modelsMeta[model] as Record<string, unknown>;
      const formFields = (result.views?.form as Record<string, unknown>)?.fields;
      if (formFields) return formFields as Record<string, unknown>;
    } catch { /* fall through */ }
    return await this.execute(model, 'fields_get', [], {
      attributes: ['string', 'type', 'relation', 'required', 'readonly'],
    }) as Record<string, unknown>;
  }

  async checkAccess(model: string, operation: string): Promise<boolean> {
    try {
      return Boolean(await this.execute(model, 'check_access_rights', [operation, false]));
    } catch { return false; }
  }

  async validFieldNames(model: string): Promise<Set<string>> {
    const meta = await this.execute(model, 'fields_get', [], { attributes: ['string'] }) as Record<string, unknown>;
    return new Set(Object.keys(meta));
  }

  async getModelId(model: string): Promise<number | null> {
    const ids = await this.execute('ir.model', 'search', [[['model', '=', model]]]) as number[];
    return ids[0] ?? null;
  }

  async userGroupXmlids(): Promise<Set<string>> {
    if (this._groupXmlids) return this._groupXmlids;
    const uid = await this.getUid();
    const rows = await this.execute('res.users', 'read', [[uid]], { fields: ['groups_id'] }) as Array<{ groups_id: number[] }>;
    const groupIds: number[] = rows[0]?.groups_id ?? [];
    const xmlids = new Set<string>();
    if (groupIds.length) {
      const imdRows = await this.execute(
        'ir.model.data', 'search_read',
        [[['model', '=', 'res.groups'], ['res_id', 'in', groupIds]]],
        { fields: ['module', 'name'] },
      ) as Array<{ module: string; name: string }>;
      for (const row of imdRows) xmlids.add(`${row.module}.${row.name}`);
    }
    this._groupXmlids = xmlids;
    return xmlids;
  }

  /** Release session state. Call on shutdown. */
  close(): void {
    this._cookies.clear();
    this._uid = null;
    this._groupXmlids = null;
  }
}
